[libpas] Add a signal for if PGM is enabled on a process

zakariaridouh opened the pull request:
#### f880a138724d2d31485d83b359f1fee597068e40
<pre>
[libpas] Add a signal for if PGM is enabled on a process
<a href="https://bugs.webkit.org/show_bug.cgi?id=287056">https://bugs.webkit.org/show_bug.cgi?id=287056</a>
<a href="https://rdar.apple.com/144202533">rdar://144202533</a>

Reviewed by Yusuke Suzuki.

This patch adds a new function to be exposed that checks if the
`pas_probabilistic_guard_malloc_can_use` value is true or false, which
is what modulates if PGM is enabled on this process during initalization
of the heap.

It also exposes this function through a JSC SPI that allows us to use
this in other places to know if PGM was enabled at all in this process,
regardless of if we made any guarded allocations or not.

This approach is better than putting additional information in the
metadata, because metadata is only generated on PGM allocations and
deallocations, which we are not guaranteed to have due to how unlikely
they are. By exposing whether or not PGM was enabled for this process
deterministically, it can help us catch bugs or crashes in the actual
implementation.

* Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
(PASPGMEnabledOnProcess):
* Source/JavaScriptCore/API/PASReportCrashPrivate.h:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_deallocate):
(pas_probabilistic_guard_malloc_enabled_on_process):
(pas_probabilistic_guard_malloc_initialize_pgm_as_enabled):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:

Canonical link: <a href="https://commits.webkit.org/293336@main">https://commits.webkit.org/293336@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90358639ebb42208584e29bd3537acaf4691d840

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98610 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18243 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103731 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49186 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18536 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26694 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75066 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32217 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101614 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14061 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89058 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55424 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13841 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7023 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48580 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91297 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83796 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/7101 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/106118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97239 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25700 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18725 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/106118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26077 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85259 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/106118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21096 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28165 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5849 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19390 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25658 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/30840 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/120857 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25476 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33821 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28796 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27051 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -67,6 +67,12 @@ static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_
+// Function to be called to check if PGM was enabled on this process, regardless of if any guarded allocations were made.
+kern_return_t pas_pgm_enabled_on_process(void)
+{
+    return pas_probabilistic_guard_malloc_can_use ? KERN_SUCCESS : KERN_FAILURE;
```
`pas_probabilistic_guard_malloc_can_use` can be set to false even if PGM had been used. See `pas_probabilistic_guard_malloc_allocate()`

zakariaridouh replied:
Great catch, I'll use a flag that clamps to `true` if `pas_probabilistic_guard_malloc_can_use` has been `true` at any point (even if it hasn't actually done any allocations)

Constellation commented on Source/JavaScriptCore/API/PASReportCrashPrivate.h:
```diff
@@ -34,6 +34,7 @@
 extern "C" {
 #endif
 
+JS_EXPORT kern_return_t PASPGMEnabledOnProcess();
```
Prototype definition for C should have void for empty parameter.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:
```diff
@@ -111,6 +111,12 @@ extern PAS_API pas_ptr_hash_map_in_flux_stash pas_pgm_hash_map_in_flux_stash;
+ * even if it been subsequently disabled, or no guarded allocations have been made
+ */
+extern PAS_API bool pas_probabilistic_guard_malloc_has_been_used;
+
```
Why does this parameter need to be exposed via header while we are offering a function getting this information?
Remove this change.

Constellation replied:
And remove this change. Now `pas_probabilistic_guard_malloc_enabled_on_process` is defined in this C file, so you do not need to expose this variable.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -67,6 +67,13 @@ static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_
+{
+    return pas_probabilistic_guard_malloc_has_been_used ? KERN_SUCCESS : KERN_FAILURE;
+}
+
```
You should define `pas_probabilistic_guard_malloc_enabled` function in `pas_probabilistic_guard_malloc_allocator.h` and `pas_probabilistic_guard_malloc_allocator.c` and use it.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -38,6 +38,7 @@
 extern "C" {
 #endif
 
+PAS_API kern_return_t pas_pgm_enabled_on_process(void);
```
Remove this.

Constellation replied:
Rename this one to `pas_probabilistic_guard_malloc_enabled_on_process`.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -51,6 +51,7 @@ static size_t pgm_metadata_index = 0;
 uint16_t pas_probabilistic_guard_malloc_random;
 uint16_t pas_probabilistic_guard_malloc_counter = 0;
 
+bool pas_probabilistic_guard_malloc_has_been_used = false;
```
You should make it `static` variable in this file and expose a function returning this information.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -67,6 +67,13 @@ static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_
 }
 
+// Function to be called to check if PGM has been enabled on this process at any point,
+// regardless of if it has since been disabled, or if any guarded allocations were made.
```
You need to use C style comment here.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.c:
```diff
@@ -67,6 +67,15 @@ static PAS_ALWAYS_INLINE kern_return_t PAS_WARN_UNUSED_RETURN pas_update_report_
+ * Function to be called to check if PGM has been enabled on this process at any point,
+ * regardless of if it has since been disabled, or if any guarded allocations were made.
+*/
+kern_return_t pas_pgm_enabled_on_process(void)
```
Please rename it to `pas_probabilistic_guard_malloc_enabled_on_process`.
And let's move this implementation to `pas_probabilistic_guard_malloc_allocator.c`.

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
```diff
@@ -53,6 +53,7 @@ uint16_t pas_probabilistic_guard_malloc_counter = 0;
 
 bool pas_probabilistic_guard_malloc_can_use = true;
 bool pas_probabilistic_guard_malloc_is_initialized = false;
+bool pas_probabilistic_guard_malloc_has_been_used = false;
```
Define it as `static`

Constellation commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -38,6 +38,7 @@
 extern "C" {
```
Why not include `pas_probabilistic_guard_malloc_allocator.h`?

Constellation reviewed (commented):
Commented, also it seems that some build failure is happening.

zakariaridouh commented on Source/bmalloc/libpas/src/libpas/pas_report_crash.h:
```diff
@@ -31,6 +31,7 @@
  * defined in JavaScriptCore PasReportCrashPrivate.h. */
 
 #ifdef __APPLE__
+#include "pas_probabilistic_guard_malloc_allocator.h"
```
not necessary here anymore, removing.

Constellation commented on Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
```diff
@@ -41,6 +42,16 @@ WTF_IGNORE_WARNINGS_IN_THIRD_PARTY_CODE_END
+#if BENABLE(LIBPAS)
+    return pas_probabilistic_guard_malloc_enabled_on_process() ? KERN_SUCCESS : KERN_FAILURE;
+#endif /* !USE(SYSTEM_MALLOC) */
+#endif /* BENABLE(LIBPAS) */
```
These two endif ordering looks incorrect

Constellation reviewed (approved):
r=me with fixes

zakariaridouh committed:
[libpas] Add a signal for if PGM is enabled on a process
https://bugs.webkit.org/show_bug.cgi?id=287056
rdar://144202533

Reviewed by Yusuke Suzuki.

This patch adds a new function to be exposed that checks if the
`pas_probabilistic_guard_malloc_can_use` value is true or false, which
is what modulates if PGM is enabled on this process during initalization
of the heap.

It also exposes this function through a JSC SPI that allows us to use
this in other places to know if PGM was enabled at all in this process,
regardless of if we made any guarded allocations or not.

This approach is better than putting additional information in the
metadata, because metadata is only generated on PGM allocations and
deallocations, which we are not guaranteed to have due to how unlikely
they are. By exposing whether or not PGM was enabled for this process
deterministically, it can help us catch bugs or crashes in the actual
implementation.

* Source/JavaScriptCore/API/PASReportCrashPrivate.cpp:
(PASPGMEnabledOnProcess):
* Source/JavaScriptCore/API/PASReportCrashPrivate.h:
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.c:
(pas_probabilistic_guard_malloc_deallocate):
(pas_probabilistic_guard_malloc_enabled_on_process):
(pas_probabilistic_guard_malloc_initialize_pgm_as_enabled):
* Source/bmalloc/libpas/src/libpas/pas_probabilistic_guard_malloc_allocator.h:

Canonical link: https://commits.webkit.org/293336@main
